coreference/bert_stuff.py

```diff
@@ -0,0 +1,175 @@
+config = pyhocon.ConfigFactory.parse_file(config_file_path)
+
+
+def generate_cdlm_embeddings_from_model(parallel_model,bert_sentences, device, batch_size=150):
+    """
+
+    Parameters
+    ----------
+    parallel_model: Parallel Model
+    bert_sentences: arr
+        list of sentences with mentions surrounded by <m> and </m>
+    device: torch.device
+    batch_size: int
+
+    Returns
+    -------
+        arr of vectors
+    """
+    parallel_model.eval()
+
+    # Store vectors here
+    all_vectors = []
+    
+    #vector_map_path = working_folder + f'/{men_type}_vectormaps.pkl'
+
+    loc_mentions = []
+    
+    #create batches of sentences or documents and get them tokenized
+    
+    batch = [bert_sentences[x] for x in range( len(bert_sentences))]
+    bert_tokens = parallel_model.module.tokenizer(batch, pad_to_max_length=True, add_special_tokens=False, truncation =False )
+    
+    input_ids = torch.tensor(bert_tokens['input_ids'], device=device)
+    attention_mask = torch.tensor(bert_tokens['attention_mask'], device=device)
+    m = input_ids.cpu()
+    
+    # get the additional special token IDs to create the binary map 
+    k = m == parallel_model.module.vals[0]
+    p = m == parallel_model.module.vals[1]
+
+    v = (k.int() + p.int()).bool()
+    nz_indexes = v.nonzero()[:, 1].reshape(m.shape[0], 2)
+    loc_mentions.extend(nz_indexes.cpu().numpy())
+
+    #get the overflowing token indices and chunk the document so that the special tokens are 
+    #within max token length(4096) for the CDLM model
+    overflow_idx=[]
+  
+    for i,j in enumerate(loc_mentions):
+        if j[1]>=4096:
+            overflow_idx.append([i, j[1] ])
+    
+    idx = [x[0] for x in overflow_idx]
+    # create a list for appending the overflowing indices
+    m_new  = []
+    for i, j in enumerate(overflow_idx):
+        bert_sent = bert_sentences[j[0] ]
+        m_new.append(m[j[0],  j[1]-4096+1:j[1]+1] )
+    #convert them into a tensor and pad to original max length of the entire batch 
+    m_new_tensor = torch.stack(m_new) 
+    m_new_tensor = F.pad(m_new_tensor,pad =(0, m.shape[1]-m_new_tensor.shape[1]), value=1) 
+
+    m[idx] = m_new_tensor
+
+    k = m == parallel_model.module.vals[0]
+    p = m == parallel_model.module.vals[1]
+
+    v = (k.int() + p.int()).bool()
+    nz_indexes = v.nonzero()[:, 1].reshape(m.shape[0], 2)
+
+    q = torch.arange(m.shape[1])
+    q = q.repeat(m.shape[0], 1)
+    
+    #create the  binary maps 
+    msk_0 = (nz_indexes[:, 0].repeat(m.shape[1], 1).transpose(0, 1)) <= q
+    msk_1 = (nz_indexes[:, 1].repeat(m.shape[1], 1).transpose(0, 1)) >= q
+
+    msk_0_ar = (nz_indexes[:, 0].repeat(m.shape[1], 1).transpose(0, 1)) < q
+    msk_1_ar = (nz_indexes[:, 1].repeat(m.shape[1], 1).transpose(0, 1)) > q
+
+    attention_mask_g = msk_0.int() * msk_1.int()
+    # get the tokens within 4096 after chunking 
+    input_ids = input_ids[:, :4096]
+    
+    attention_mask[:, 0] = 2
```

Comment:
what does 2 mean?

Reply:
attention_mask[:, 0] = 1 # 1 because we are letting this token to be attended globally, not 2 because longformer issue resolved https://github.com/huggingface/transformers/issues/7015